### DISCUSSION TITLE
Full 2-port simulation (Python)

### DISCUSSION THREAD

**Discussion by VolkerMuehlhaus** [thliebig/openEMS-Project#64] [Q&A Usage] [answered]
Hi all,

is there any example how to do full 2-port simulation with **excitation at port 1 and 2 (one after another)** using the **Python** interface?
For Matlab I found the directional_coupler example, but the syntax there is very tricky and I have no idea how to implement this using Python interface. Anyone any example?

What I tried: set port 1 excitation first and run FDTD, then swap port1.excite and port2.excite values using the same (!) CSX and re-run FDTD, but that seems to be the wrong approach. 

All help/hint/example is much appreciated!
Volker

**Comment by 0xCoto** [comment]
That's also how I do it: excite the port of interest and set the others to excite=0.0.

**Reply by VolkerMuehlhaus** [reply]
Could you share the code how you modify excitation after the initial run?

My task is to get the full [S] matrix, so I need one run with port 1 excited and then another run with port 2 excited, in differnet directories.

Exciting port 1 first, with no excitation at port 2, is done when creating the port instances. But when that initial FDTD simulation is finished, how to set change port excitations for the reverse path? It seems to me that I can't just set port.excite to different values and call FDTD.Run() again?

**Reply by 0xCoto** [reply]
I do something a bit more complicated because I'd run simulations on the cloud, so I have set up a custom orchestrator to accelerate the speed of the simulations as much as possible. I do this by delegating each port excitation to a separate server instance/node before merging the results of the S-matrix.

**Comment by VolkerMuehlhaus** [comment, marked as answer]
Ok, maybe I understand the concept better now: it seems that we need a separate, new CSX for each separate port excitation. My solution was to place the entire code that populates the CSX into a function, and call that function two times with varying settings for port excitation. Each excitation data is placed in a separate directory, so we can finally get the full [S] matrix including reverse path.

Just in case another beginner has the same question, I tried to post some code snippets from my Python model, but the code embedding feature messed up formatting of my code.

Marking thread as solved.

Best regards
Volker

Function to create CSX, port excitation (forward/reverse path of 2-port) set by parameter:

```python

def createSimulation (exciteport):
# Define function for model creation because we need to create and run separate CSX
# for each excitation. For S11,S21 we only need to excite port 1, but for S22,S12
# we need to excite port 2. This requires separate CSX with different port settings.


    ############ Geometry setup ############
    CSX = ContinuousStructure()
    FDTD.SetCSX(CSX)
    mesh = CSX.GetGrid()
    mesh.SetDeltaUnit(unit)

(...)

    if exciteport==1:
        port1 = FDTD.AddLumpedPort(1, 50, [-22.2, 0, TopMetal1_zmin], [-10, 10, TopMetal1_zmax], 'x', excite=1.0, priority=150)
        port2 = FDTD.AddLumpedPort(2, 50, [ 22.2, 0, TopMetal1_zmin], [ 10, 10, TopMetal1_zmax], 'x', excite=0,   priority=150)
    else:    
        port1 = FDTD.AddLumpedPort(1, 50, [-22.2, 0, TopMetal1_zmin], [-10, 10, TopMetal1_zmax], 'x', excite=0,   priority=150)
        port2 = FDTD.AddLumpedPort(2, 50, [ 22.2, 0, TopMetal1_zmin], [ 10, 10, TopMetal1_zmax], 'x', excite=1.0, priority=150)

(...)
    
    # create subdirectory to hold data for this excitation
    excitation_path = os.path.join(sim_path, 'sub-' + str(exciteport))
    if not os.path.exists(excitation_path):
        os.makedirs(excitation_path)

    # write CSX file 
    CSX_file = os.path.join(excitation_path, model_basename + '.xml')
    CSX.Write2XML(CSX_file)

    if not postprocess_only: # preview model, but only for first port excitation
        if (exciteport==1): 
            from CSXCAD import AppCSXCAD_BIN
            os.system(AppCSXCAD_BIN + ' "{}"'.format(CSX_file))

    if not preview_only:  # start simulation 
        if not postprocess_only:
            print("Running FDTD simulation, excitation port " + str(exciteport))
            FDTD.Run(excitation_path, verbose=1)


    # return ports, so that we can postprocess them
    return port1, port2, excitation_path
  
######### end of function createSimulation (exciteport) ##########
``` 

This createSimulation () function is called in main program, to create two CSX in separate directories and run each of them:

```python
########### create model, run and post-process ###########

f = np.linspace(fstart,fstop,numfreq)

# call createSimulation function defined above 
sub1_port1, sub1_port2, sub1_excitation_path = createSimulation (1)  # excite port 1 
sub2_port1, sub2_port2, sub2_excitation_path = createSimulation (2)  # excite port 2

if not preview_only:
    # evaluate port 1 excitation
    sub1_port1.CalcPort( sub1_excitation_path, f, ref_impedance = 50)
    sub1_port2.CalcPort( sub1_excitation_path, f, ref_impedance = 50)

    s11 = sub1_port1.uf_ref / sub1_port1.uf_inc
    s21 = sub1_port2.uf_ref / sub1_port1.uf_inc

    # evaluate port 2 excitation
    sub2_port1.CalcPort( sub2_excitation_path, f, ref_impedance = 50)
    sub2_port2.CalcPort( sub2_excitation_path, f, ref_impedance = 50)

    s22 = sub2_port2.uf_ref / sub2_port2.uf_inc
    s12 = sub2_port1.uf_ref / sub2_port2.uf_inc

```

**Reply by thliebig** [reply]
Hello Volker, yes this should work, but I will have a look why it is not possible to just enable/disable a port and re-simulate (in a new folder)...

br
Thorsten

**Reply by RFLAH** [reply]
Hello,

I was able add multi-port excitation up to 8 ports for now with help of Volker's instructions. It was a lot of copy paste in the code setting up the identical primitives for different CSXs and simulation paths. It would be great to have just one CSX and mesh for multi-port excitation and excitation port enable/disable function as Torsten wrote. 

Before this I tried pretty much everything I could think of to try to edit the port excitation parameters without any luck - the original port parameters always appeared in the XML. I thought about brute force parsing and manually editing the XML-file before running the next excitation, but it would have taken some effort and I am not sure if it would have worked anyway. I was also looking for an option to delete the already excited port from the CSX? and and adding it again with 0 excite value before running the next excitation, but I could not find support for this from current API.

I am currently using my own GUI to setup the primitives, mesh and ports, but I would prefer using the CSXCAD interface at least for 3D modelling more complex and parameterized models and maybe meshing with possibility to save/load models easily. My next step is probably to try, if I could get the multi-port simulations also working with primitives created fully on CSXCAD side and using ReadFromXML to different excitation CSXs since one can only edit one (CSX) XML at the time in CSXCAD, which also supports the idea of single CSX use.

I am using CSXCAD for mainly previewing the model and mesh, but I would like to use it more at least for building the models and complement the missing CSXCAD API features with my separate Python API GUI, which maybe will become needless on some day as CSXCAD gets more features implemented.

**Reply by thliebig** [reply]
Well I think the redundancy should really not be necessary. Why not use a variable for the excite keywords? e.g. an array/list where the port number (-1) is the array index? Then every port needs only to be defined once and you can loop through the different simulations with different ports set active???

For two ports Volkers solution is good enough, but more than that I would e.g. use a list or dict instead...

**Reply by VolkerMuehlhaus** [reply]
@thliebig Sounds interesting, would you have time to modify my code snippet above for list or dict?

**Reply by thliebig** [reply]
In your two port example you would not even need a list, just something like this maybe:
```python
        port1 = FDTD.AddLumpedPort(1, 50, [-22.2, 0, TopMetal1_zmin], [-10, 10, TopMetal1_zmax], 'x', excite=float(exciteport==1), priority=150)
        port2 = FDTD.AddLumpedPort(2, 50, [ 22.2, 0, TopMetal1_zmin], [ 10, 10, TopMetal1_zmax], 'x', excite=float(exciteport==2),   priority=150)
```
But this would also work for many more ports...

**Comment by aholtzma-am** [comment]
Changing the port excitation doesn't work as CSX.AddExcitation() is only called in ```__init__()```. Each of the port types needs a set_excitation() handler to handle changes after init.

**Reply by VolkerMuehlhaus** [reply]
@thliebig It would really be great to get this fixed, and be able to switch excitation after creating the FDTD API call, like we can do in the Matlab API. I am having real trouble to run multiple port excitations on the same model in my Python code.

In theory I know how to do this, by just re-building the entire FDTD again for each excitation, but the software behaves unpredictable sometimes. Runs only 2 or 3 out of 4 excitations, or runs the first excitation and then throws errors like 

```
 File "openEMS/openEMS.pyx", line 491, in openEMS.openEMS.openEMS._SetLibraryArguments
RuntimeError: option '--verbose=1' is ambiguous and matches different versions of '--verbose'
```
I'm really frustrated. The code behaves different on Ubuntu than Windows etc ... all very strange. It would be so nice to just switch excitations from the SAME model, like we can do in Matlab.

**Reply by thliebig** [reply]
Hi, the Windows version is still quite old as I do not rebuilt it often as this is really a hassle on Windows.
If you use always the latest Linux Version, than for sure the versions will (currently) not behave the same.

Can you specify what does not work for you? Enabling/Disabling the "excite" flag on creation should work right?

**Reply by thliebig** [reply]
I had a look, there is currently no way to deactivate an already defined (and excited) port. Since the python interface is much more versatile, this is clearly a missing feature. I will have a look into it.

**Reply by VolkerMuehlhaus** [reply]
@thliebig Thanks for your quick reply, Thorsten! 

Yes, I can create a model and set the excitation at that moment, but there is no way to enable/disable the excitations after running simulation, to re-run the reverse path of an already create two-port model. If you go back to the beginning of this thread, you hopefully understand what I mean.

So I am using a workaround, and generate a "new" FDTD for each excitation, which works on Windows. When running that exact model on Ubuntu, I just got the funny effect that the first excitation runs and the second gives this totally unexpected error about solver parameters:

```
[@     5m34s] Timestep:       627689 || Speed:  601.9 MC/s (5.381e-04 s/TS) || Energy: ~2.84e-23 (-59.79dB)
[@     5m39s] Timestep:       635912 || Speed:  562.7 MC/s (5.756e-04 s/TS) || Energy: ~2.54e-23 (-60.27dB)
Time for 635912 iterations with 323901.00 cells : 339.11 sec
Speed: 607.39 MCells/s 
FDTD simulation completed successfully.

________________________
Mesh cells by axis (total 324 kcells):
 x = 87
 y = 51
 z = 73
Smallest cell size:
 dx = 0.4950
 dy = 0.5000
 dz = 0.1000
________________________

Starting FDTD simulation
Traceback (most recent call last):
  File "/home/volker/openems_work/work/run_rfcmim_2port_full.py", line 106, in <module>
    data_paths.append(simulation_setup.runSimulation (excite_ports, FDTD, sim_path, model_basename, preview_only, postprocess_only))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/volker/openems_work/work/modules/util_simulation_setup.py", line 284, in runSimulation
    FDTD.Run(excitation_path, verbose=1)
  File "openEMS/openEMS.pyx", line 530, in openEMS.openEMS.openEMS.Run
  File "openEMS/openEMS.pyx", line 491, in openEMS.openEMS.openEMS._SetLibraryArguments
RuntimeError: option '--verbose=1' is ambiguous and matches different versions of '--verbose'
```

I have some wrappers around core openEMS to simplify the workflow for IHP RFIC EM users, so the model code will look strange to you, but essentially it's creating the same model just with excitation at port 1, then excitation at port 2.

```
# Create simulation for port 1 and 2 excitation, return value is list of data paths, one for each excitation
data_paths = []
for excite_ports in [[1],[2]]:  # list of ports that are excited one after another
    FDTD = openEMS(EndCriteria=exp(energy_limit/10 * log(10)))
    FDTD.SetGaussExcite( (fstart+fstop)/2, (fstop-fstart)/2 )
    FDTD.SetBoundaryCond( Boundaries )
    FDTD = simulation_setup.setupSimulation (excite_ports, simulation_ports, FDTD, materials_list, dielectrics_list, ...)
    data_paths.append(simulation_setup.runSimulation (excite_ports, FDTD, sim_path, model_basename, preview_only, postprocess_only))

```

The underlying solver call is very basic:

```
    if not (preview_only or postprocess_only):  # start simulation 
        print('Starting FDTD simulation')
        try:
            FDTD.Run(excitation_path, verbose=1)
            print('FDTD simulation completed successfully.')
        except AssertionError as e:
            print('[ERROR] AssertionError during FDTD simulation: ', e)
            sys.exit(1)

```

So we have the exact same calls, just the excited ports differ. Runs fine on Win, fails on Ubuntu with the message above. 
It would be great if my workaround for iterated excitations would no longer be necessary.

**Reply by thliebig** [reply]
Can you tell me which version on Linux you use?
I think currently it should be "v0.0.36-63", that is 63 changes after 0.0.36 ;)

> openEMS 64bit -- version v0.0.36-63-gc2247a4

because this "_SetLibraryArguments" ambiguous sounds familiar and I believe was fixed? Not sure, I will test.

**Reply by thliebig** [reply]
Nope, I can confirm the issue...

**Reply by thliebig** [reply]
But this only happens **if** you set verbose? If you do not, everything runs fine. Not a fix but a work-around for now?

**Reply by VolkerMuehlhaus** [reply]
> But this only happens **if** you set verbose? If you do not, everything runs fine. Not a fix but a work-around for now?

Thanks Thorsten, this workaround works, fine for me!

**Comment by thliebig** [comment]
The latest version (build from source) now has many new options for this. Ports can now be enabled/disabled (SetEnabled), but the correct "excite" must be set for the ports on setup.
Example:


```
for active_port in [1,2]:
    port_1.SetEnabled(active_port == 1)
    port_2.SetEnabled(active_port == 2)
    _sim_path = os.path.join(Sim_Path, 'port-{}'.format(active_port))
    if not post_proc_only:
        FDTD.Run(_sim_path, cleanup=True)

    ### Post-processing and plotting
    f = linspace( 1e6, f_max, 1601 )
    port_1.CalcPort( _sim_path, f, ref_impedance = 50)
    port_2.CalcPort( _sim_path, f, ref_impedance = 50)

    if active_port==1:
        s11 = port_1.uf_ref / port_1.uf_inc
        s21 = port_2.uf_ref / port_1.uf_inc
    else:
        s22 = port_2.uf_ref / port_2.uf_inc
        s12 = port_1.uf_ref / port_2.uf_inc
```

But it is now also possible to reset everything. Either the FDTD class or only the geometry using "Reset()".
What's an open issue is to create multiple instances of `FDTD = openEMS()` as this will cause the library arguments traceback...

**Reply by VolkerMuehlhaus** [reply]
Many thanks Thorsten, much appreciated! This will simplify things a lot in the future!

> What's an open issue is to create multiple instances of `FDTD = openEMS()` as this will cause the library arguments traceback...

Ok, now I'm worried, because I still need to create a solution that works on Windows as well. What would be the correct sequence of things to create one simulation for port 1 excitation, then another simulation for port 2 excitation using the **old** version? 

Background is that I'm creating a reusable workflow for other users, and I don't know if they will be running in Windows or Linux. I will send you an email, maybe you have time to have a look.
